fem/fe.cpp

```diff
@@ -3272,6 +3272,203 @@ const
+{
+   Nodes.IntPoint(0).x = 0.0;
+   Nodes.IntPoint(0).y = 0.0;
+   Nodes.IntPoint(0).z = 0.0;
+   Nodes.IntPoint(1).x = 1.0;
+   Nodes.IntPoint(1).y = 0.0;
+   Nodes.IntPoint(1).z = 0.0;
+   Nodes.IntPoint(2).x = 0.0;
+   Nodes.IntPoint(2).y = 1.0;
+   Nodes.IntPoint(2).z = 0.0;
+   Nodes.IntPoint(3).x = 0.0;
+   Nodes.IntPoint(3).y = 0.0;
+   Nodes.IntPoint(3).z = 1.0;
+   Nodes.IntPoint(4).x = 1.0;
+   Nodes.IntPoint(4).y = 0.0;
+   Nodes.IntPoint(4).z = 1.0;
+   Nodes.IntPoint(5).x = 0.0;
+   Nodes.IntPoint(5).y = 1.0;
+   Nodes.IntPoint(5).z = 1.0;
+}
+
+void LinearWedgeFiniteElement::CalcShape(const IntegrationPoint &ip,
+                                         Vector &shape) const
+{
+   shape(0) = (1. - ip.x - ip.y) * (1. - ip.z);
+   shape(1) = ip.x * (1. - ip.z);
+   shape(2) = ip.y * (1. - ip.z);
+   shape(3) = (1. - ip.x - ip.y) * ip.z;
+   shape(4) = ip.x * ip.z;
+   shape(5) = ip.y * ip.z;
+}
+
+void LinearWedgeFiniteElement::CalcDShape(const IntegrationPoint &ip,
+                                          DenseMatrix &dshape) const
+{
+   dshape(0,0) = -1. + ip.z;
+   dshape(0,1) = -1. + ip.z;
+   dshape(0,2) = -1. + ip.x + ip.y;
+
+   dshape(1,0) =  1. - ip.z;
+   dshape(1,1) =  0.;
+   dshape(1,2) = -ip.x;
+
+   dshape(2,0) =  0.;
+   dshape(2,1) =  1. - ip.z;
+   dshape(2,2) = -ip.y;
+
+   dshape(3,0) = -ip.z;
+   dshape(3,1) = -ip.z;
+   dshape(3,2) =  1. - ip.x - ip.y;
+
+   dshape(4,0) =  ip.z;
+   dshape(4,1) =  0.;
+   dshape(4,2) =  ip.x;
+
+   dshape(5,0) =  0.;
+   dshape(5,1) =  ip.z;
+   dshape(5,2) =  ip.y;
+}
+
+void LinearWedgeFiniteElement::GetFaceDofs (int face, int **dofs, int *ndofs)
+const
+{
+   static int face_dofs[5][4] =
+   {{0, 2, 1, -1}, {3, 4, 5, -1}, {0, 1, 4, 3}, {1, 2, 5, 4}, {2, 0, 3, 5}};
+
+   *ndofs = (face < 2) ? 3 : 4;
+   *dofs  = face_dofs[face];
+}
+
+
+LinearPyramidFiniteElement::LinearPyramidFiniteElement()
+   : NodalFiniteElement(3, Geometry::PYRAMID, 5, 1)
+{
+   Nodes.IntPoint(0).x = 0.0;
+   Nodes.IntPoint(0).y = 0.0;
+   Nodes.IntPoint(0).z = 0.0;
+   Nodes.IntPoint(1).x = 1.0;
+   Nodes.IntPoint(1).y = 0.0;
+   Nodes.IntPoint(1).z = 0.0;
+   Nodes.IntPoint(2).x = 1.0;
+   Nodes.IntPoint(2).y = 1.0;
+   Nodes.IntPoint(2).z = 0.0;
+   Nodes.IntPoint(3).x = 0.0;
+   Nodes.IntPoint(3).y = 1.0;
+   Nodes.IntPoint(3).z = 0.0;
+   Nodes.IntPoint(4).x = 0.0;
+   Nodes.IntPoint(4).y = 0.0;
+   Nodes.IntPoint(4).z = 1.0;
+}
+
+void LinearPyramidFiniteElement::CalcShape(const IntegrationPoint &ip,
+                                           Vector &shape) const
+{
+   double x = ip.x, y = ip.y, z = ip.z;
+   double ox = 1.-x-z, oy = 1.-y-z, oz = 1.-z;
+
+   double tol = 1e-6;
+
+   if (oz <= tol)
+   {
+      // We must return the limit of the basis functions as z->1.  In
+      // order to remain inside the pyramid in this limit the x and y
+      // coordinates must be approaching 0. The resulting limiting
+      // basis function values are:
```

Comment:
For consistency with the rest of mfem, can you please consider aligning the comments and elsewhere to 80 columns?

For example the above can fit in 3 lines:

```c++
   if (oz <= tol)
   {
      // We must return the limit of the basis functions as z->1. In order to
      // remain inside the pyramid in this limit the x and y coordinates must be
      // approaching 0. The resulting limiting basis function values are:

      shape(0) = 0.;
      shape(1) = 0.;
      ...
```